### PR TITLE
Use custom DataType for count aggregation to avoid boxing long to Long

### PR DESCRIPTION
**Pull request by matriv**

### CODE REVIEW

**Comment by dobe** [comment]
pls add changelog entry about count optimization,
rebase

**Comment by dobe** [comment]
ack

File: sql/src/main/java/io/crate/operation/projectors/ProjectionToProjectorVisitor.java [hunk 1]
```diff
@@ -167,6 +167,12 @@ public Projector visitGroupProjection(GroupProjection projection, Context contex
     }
 
     @Override
+    public Projector visitMergeCountProjection(MergeCountProjection projection, Context context) {
+        ImplementationSymbolVisitor.Context symbolContext = new ImplementationSymbolVisitor.Context();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/projectors/ProjectionToProjectorVisitor.java]: why are you creating a context that isn't used?

File: sql/src/main/java/io/crate/planner/projection/MergeCountProjection.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import io.crate.types.DataTypes;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+import java.util.List;
+
+public class MergeCountProjection extends Projection {
+
+    public static final MergeCountProjection PARTIAL_COUNT_AGGREGATION_PROJECTION = new MergeCountProjection();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/projection/MergeCountProjection.java]: I think the name doesn't fit?

File: sql/src/main/java/io/crate/planner/projection/MergeCountProjection.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+    public MergeCountProjection() {
+    }
+
+    @Override
+    public List<? extends Symbol> outputs() {
+        return OUTPUTS;
+    }
+
+    public Symbol value() {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/projection/MergeCountProjection.java]: where is this used?

File: sql/src/main/java/io/crate/planner/projection/MergeCountProjection.java (hunk 2 above)
Comment (matriv) [inline sql/src/main/java/io/crate/planner/projection/MergeCountProjection.java]: I kept it as it was before... Maybe MERGE_COUNT_AGGREGATION_PROJECTION or any other suggestion?

File: sql/src/main/java/io/crate/planner/projection/MergeCountProjection.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        Symbol.toStream(value, out);
+    }
+
+    @Override
+    public RowGranularity requiredGranularity() {
+        return RowGranularity.SHARD;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/projection/MergeCountProjection.java]: I think this is should be cluster and not shard

**Comment by mfussenegger** [comment]
lgtm

**Pull request merged**